mypy/checkexpr.py

```diff
@@ -1130,22 +1135,15 @@ def check_overload_call(self,
         unioned_result = None  # type: Optional[Tuple[Type, Type]]
         unioned_errors = None  # type: Optional[MessageBuilder]
         union_success = False
-        if any(isinstance(arg, UnionType) and len(arg.relevant_items()) > 1  # "real" union
-               for arg in arg_types):
-            erased_targets = self.overload_erased_call_targets(plausible_targets, arg_types,
-                                                               arg_kinds, arg_names, context)
-            unioned_callable = self.union_overload_matches(erased_targets)
-
-            if unioned_callable is not None:
-                unioned_errors = arg_messages.clean_copy()
-                unioned_result = self.check_call(unioned_callable, args, arg_kinds,
-                                                 context, arg_names,
-                                                 arg_messages=unioned_errors,
-                                                 callable_name=callable_name,
-                                                 object_type=object_type)
-                # Record if we succeeded. Next we need to see if maybe normal procedure
-                # gives a narrower type.
-                union_success = unioned_result is not None and not unioned_errors.is_errors()
+        if any(self.real_union(arg) for arg in arg_types):
+            unioned_errors = arg_messages.clean_copy()
+            unioned_result = self.union_overload_result(plausible_targets, args, arg_types,
+                                                        arg_kinds, arg_names,
+                                                        callable_name, object_type,
+                                                        context, arg_messages=unioned_errors)
+            # Record if we succeeded. Next we need to see if maybe normal procedure
+            # gives a narrower type.
+            union_success = unioned_result is not None and not unioned_errors.is_errors()
```

Comment:
Given the way you implemented `union_overload_result`, I think it's impossible for `union_errors` to actually ever contain an error. As a consequence, `union_success` will be true if and only if `union_result is not None` is also true.

In that case, I think we can simplify some of the logic by getting rid of `union_success` and using just `union_result`. E.g. here, we replace this line with `if union_errors.is_errors(): unioned_result = None` just in case and get rid of `union_success` entirely.